keps/sig-network/1024-nodelocal-cache-dns/README.md

```diff
@@ -1,68 +1,74 @@
+    - [iptables NOTRACK](#iptables-notrack)
+    - [Choice of caching agent](#choice-of-caching-agent)
+    - [Metrics](#metrics)
   - [Risks and Mitigations](#risks-and-mitigations)
-- [Design Details](#design-details)
-  - [Test Plan](#test-plan)
-  - [Graduation Criteria](#graduation-criteria)
-  - [Alternatives](#alternatives)
+- [Graduation Criteria](#graduation-criteria)
+- [Rollout Plan](#rollout-plan)
 - [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
 <!-- /toc -->
 
 ## Summary
 
-[NodeLocal DNSCache](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/nodelocaldns/README.md) is an addon that runs a dnsCache pod as a daemonset to improve clusterDNS performance and reliability. The feature has been in Alpha since 1.13 release. This document lays out the plan to promote it to beta.
+This proposal aims to improve DNS performance by running a dns caching agent on cluster nodes as a Daemonset. In today's architecture, pods in ClusterFirst DNS mode reach out to a kube-dns serviceIP for DNS queries. This is translated to a kube-dns endpoint via iptables rules added by kube-proxy. With this new architecture, pods will reach out to the dns caching agent running on the same node, thereby avoiding iptables DNAT rules and connection tracking. The local caching agent will query kube-dns for cache misses of cluster hostnames(cluster.local suffix by default).
+
 
 ## Motivation
 
-NodeLocal DNSCache has been in Alpha for the past 2 releases and users have deployed it to fix DNS performance issues.
-- [Example 1](https://github.com/kubernetes/kubernetes/issues/56903#issuecomment-485353223)
-- [Example 2](https://github.com/kubernetes/kubernetes/issues/45363#issuecomment-443019910)
+* With the current DNS architecture, it is possible that pods with the highest DNS QPS have to reach out to a different node, if there is no local kube-dns instance.  
+Having a local cache will help improve the latency in such scenarios. 
 
-Based on the feedback so far, we feel the feature is ready to be graduated to beta.
+* Skipping iptables DNAT and connection tracking will help reduce [conntrack races](https://github.com/kubernetes/kubernetes/issues/56903) and avoid UDP DNS entries filling up conntrack table.
 
-### Goals
+* Connections from local caching agent to kube-dns can be upgraded to TCP. TCP conntrack entries will be removed on connection close in contrast with UDP entries that have to timeout ([default](https://www.kernel.org/doc/Documentation/networking/nf_conntrack-sysctl.txt) `nf_conntrack_udp_timeout` is 30 seconds)
 
-Graduate NodeLocal DNSCache to beta.
+* Upgrading DNS queries from UDP to TCP would reduce tail latency attributed to dropped UDP packets and DNS timeouts usually up to 30s (3 retries + 10s timeout). Since the nodelocal cache listens for UDP DNS queries, applications don't need to be changed.
```

Comment:
isn't 5s the default timeout per query? Is 3 retries a standard application retry? Asking since this looks like newly added content, not copy-paste.

Reply:
It's actually directly copied from the existing KEP - see 4th bullet here:
https://github.com/kubernetes/enhancements/blob/13a4bd1c2eb29d39275ba433ecf952882e0092c5/keps/sig-network/0030-nodelocal-dns-cache.md#motivation

[I'm happy to change if you suggest the expected version]

Reply:
Thanks! I think this is because of 5s for A and 5s for AAAA lookup.  Maybe change it to "(3 retries * (5s timeout for A + 5s timeout for AAAA))

Reply:
done